HPCC-18507 Serialize base64 binaries to/from json.

jakesmith opened the pull request:
Serializing property trees to JSON, then deserializing them back
again cause binary information to effectively be corrupted.
The notion that the value was encoded (base64) was lost and
a getPropBin call would not decode the binary.

Fix by introducing a JSON convention of $<prop> name for binaries.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

Passed regression suite.
Passed test in JIRA.

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jakesmith commented:
@afishbeck - please review

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: eeadde8e32ea2717b2654f3557b6a1bcf8df4880
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 41.664 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |92 |92 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |744 |744 |0 |
|test (thor) |639 |639 |0 |
|test (roxie) |772 |772 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented on system/jlib/jptree.cpp:
```diff
@@ -5734,19 +5737,25 @@ static void _toJSON(const IPropertyTree *tree, IIOStream &out, unsigned indent,
-            writeCharToStream(out, '[');
-            repeatingElement = true;
-            delimit = false;
+            match = (isBinary==sub->query().isBinary(NULL)) && streq(name, sub->query().queryName());
```
Not that it would be common, but what happens if the same element is added multiple times, but sometimes binary and sometimes not?
I.e.

```
tree->addPropBin("abc",mb.length(), mb.length(), mb.toByteArray());
tree->addProp("abc", "xyz";
tree->addPropBin("abc",mb.length(), mb.length(), mb.toByteArray());
tree->addProp("abc", "xyz";
tree->addPropBin("abc",mb.length(), mb.length(), mb.toByteArray());
tree->addProp("abc", "xyz";
```
I would think the proper JSON would be 2 arrays {"$abc": ["<data>", "<data>", "<data>"], "abc": ["xyz", "xyz", "xyz"]}

But unless the binaries end up grouped together, would this current implementation end up in 6 single entries (or if some are grouped a variable number of entries)?

jakesmith replied:
Items will be grouped in JSON output as per your example, if they are consecutive. 
But don't think it should otherwise, because that would lose order which is normally preserved for matching property names.

afishbeck replied:
I understand wanting to preserve the order.  It technically would be invalid JSON unless you further mangled the names to make each unique.  Loading it back into a PropertyTree would still work though.

jakesmith replied:
Right I see, didn't realize it was invalid JSON, that's unfortunate.
It would be pain to preserve order in more convoluted way , but without it, it breaks existing ordering semantics.

Any suggestions?

afishbeck replied:
You could detect this case and simply further mangle the name.  I.e. something like:
```
{"a": {"b$": ["<data>", "<data>"], "b": ["abc"], "b$.2": ["<data>"], "b.2": ["abc"]}}
```
It's a complicated work around for a rare issue, but we could standardize on it in case similar issues came up in the future... always treat trailing ".n" as a continuation of a repeating element.

Certainly open to other variations of name mangling, or any other ideas.

afishbeck commented:
@jakesmith question about mixed binary and non-binary.

jakesmith commented:
@afishbeck - added a reply.

afishbeck commented:
@jakesmith added follow on comment.

jakesmith committed:
HPCC-18548 Ensure JSON #value's are deserialized correctly

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

jakesmith committed:
HPCC-18193 Serialize base64 binaries to/from json.

Serializing property trees to JSON, then deserializing them back
again cause binary information to effectively be corrupted.
The notion that the value was encoded (base64) was lost and
a getPropBin call would not decode the binary.

Fix by Introducing a #valuebin to 'complex' type to denote
the binary value.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

jakesmith commented:
@afishbeck - please review last commit.

NB: this is based on the as yet unmerged HPCC-18548

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 8b9ca60349bbf22db78df0f9b9a460b169d7dd69
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 43.881 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |92 |92 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |744 |744 |0 |
|test (thor) |639 |639 |0 |
|test (roxie) |772 |772 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented on system/jlib/jptree.cpp:
```diff
@@ -6635,13 +6635,18 @@ class CJSONReader : public CJSONReaderBase<X>, implements IPTreeReader
             return;
         }
-        else if ('#'==*name)
+        else if ('#'==*name) // binary
```
comment is a bit misleading, could be binary and/or mixed content.

jakesmith replied:
yes, a hangover from previous version, I'll remove.

afishbeck commented on system/jlib/jptree.cpp:
```diff
@@ -6635,13 +6635,18 @@ class CJSONReader : public CJSONReaderBase<X>, implements IPTreeReader
             if (iptValue)
             {
-                retValue->swapWith(value);
+                *isValueBinary = 0 == strncmp(name+6, "bin", 3);
```
would it be more explicit (i.e. not match "#valuebinabc") to just do?:

`*isValueBinary = streq(name+6, "bin");`

jakesmith replied:
agree, will change.

afishbeck reviewed (commented):
@jakesmith minor comments, otherwise looks good.

jakesmith committed:
HPCC-18193 Minor changes following review.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

jakesmith commented:
@afishbeck - please review new commit.

afishbeck reviewed (approved):
@jakesmith looks good

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: d43d46d156bad5fc8273c3fd30b9f1ce2ecc3e27
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 45.153 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |92 |92 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |744 |744 |0 |
|test (thor) |639 |639 |0 |
|test (roxie) |772 |772 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
@jakesmith This seems to reference the wrong Jira - can you find the right one and at least get the link correct in one direction (too late to change commit message now)

jakesmith commented:
hm, yes it should have been HPCC-18507. I've attached the PR to the issue.
@richardkchapman - shall I change PR title too?
